### PR TITLE
Adding Preseed modifying capabilities to role

### PR DESCRIPTION
**Pull request by BaptisteGerondeau**
This patch adds capabilities to modify preseed.
It also changes some defaults (public and known_good should be false), and by default gives '/dev/null' as a file.

Currently no content is returned when no file is given and preseed exists on the MrP server.

### CODE REVIEW

**Comment by danrue** [comment]
Can you explain why you believe the defaults for pubic and known_good should be false?

**Comment by BaptisteGerondeau** [comment]
In my opinion you should put those flags to true after having made sure that the preseed goes all the way. So either you have the knowledge beforehand, in which case you explicitly force those to true in the playbook, or you are testing, in which case if it succeeds you can go through the interface, or modify the preseed via the role.
But if you object to that I'll happily revert everything back to True by default.

**Comment by niedbalski** [comment]
@BaptisteGerondeau in my opinion there is no a better default, but I could argue (as operator)
that you have the knowledge beforehand that the preseed  you are uploading will work (and if not, the role should provide you a way to replace it, indeed).

My suggestion is for you to keep the defaults as they are now , so we can focus on the
relevant part of the code you are proposing.

File: library/mr_provisioner_preseed.py [hunk 1]
```diff
@@ -77,74 +80,153 @@
...
+            if preseed['name'] == self.name:
+                self.id = preseed['id']
+                return True
+
+        return False
+
+    def _get_preseed_from_file(self):
+        json_preseed = {}
+
+        f = open(self.file, 'r')
```
Comment (niedbalski) [inline library/mr_provisioner_preseed.py]: with open(self.file, 'r') as fd: 
    lines = fd.readlines()

File: library/mr_provisioner_preseed.py [hunk 2]
```diff
@@ -77,74 +80,153 @@
...
+        modify (+ id). Maybe implement a jinja2 syntax check ? But that should
+        be done on mrp's side"""
+        res = {}
+        try:
+            exists = self._check_for_existence()
+        except ProvisionerError as e:
+            res['error'] = str(e)
+            return res
+
+        if not exists and self.file == '/dev/null':
```
Comment (niedbalski) [inline library/mr_provisioner_preseed.py]: Why using /dev/null as a magic string instead of '' ?

File: library/mr_provisioner_preseed.py [hunk 3]
```diff
@@ -77,74 +80,153 @@
...
+            res['error'] = 'Preseed does not exist and file not given'
+            return res
+
+        if(self.id != -1 and self.file != '/dev/null'):     #Exists and file given
```
Comment (niedbalski) [inline library/mr_provisioner_preseed.py]: Why using /dev/null as a magic string instead of '' ? See above.

File: library/mr_provisioner_preseed.py [hunk 4]
```diff
@@ -77,74 +80,153 @@
...
+        if(self.id == -1):
+            raise ProvisionerError('preseed ID is undefined, please use upload_preseed')
+        url_id = '/api/v1/preseed/' + str(self.id)
+        url = urljoin(self.url, url_id)
+        preseed = self._get_preseed_from_file()
+        r = requests.put(url, headers=self.authhead, data=json.dumps(preseed))
+        if r.status_code != 200:
+            raise ProvisionerError('Error putting preseed {} at ID {}, \
+                                   HTTP {} {}'.format(self.name, self.id, r.status_code, r.reason))
+        else:
```
Comment (niedbalski) [inline library/mr_provisioner_preseed.py]: No need to else: , you can just return

File: library/mr_provisioner_preseed.py [hunk 5]
```diff
@@ -77,74 +80,153 @@
...
-        known_good=dict(type='bool', required=False, default=False),
-        public=dict(type='bool', required=False, default=False),
+        description = dict(type='str', required=False, default=''),
+        name = dict(type='str', required=True),
+        type = dict(type='str', required=False, default='preseed'),
+        path = dict(type='str', required=False, default='/dev/null'),
+        url = dict(type='str', required=True),
+        token = dict(type='str', required=True),
+        known_good = dict(type='bool', required=False, default=False),
+        public = dict(type='bool', required=False, default=False),
```
Comment (niedbalski) [inline library/mr_provisioner_preseed.py]: Is the space required by flake8 or pylint?

File: tasks/provision.yml [hunk 6]
```diff
@@ -42,7 +42,7 @@
     name: "{{ mr_provisioner_preseed_name }}"
     description: "{{ mr_provisioner_preseed_description | default('') }}"
     type: "{{ mr_provisioner_preseed_type | default('preseed')}}"
-    path: "{{ mr_provisioner_preseed_path }}"
+    path: "{{ mr_provisioner_preseed_path | default('/dev/null')}}"
```
Comment (niedbalski) [inline tasks/provision.yml]: Why using /dev/null as a magic string instead of '' ? See above.

**Review by niedbalski — CHANGES_REQUESTED** [review]
Thank you for rebasing your changes into a single commit. 
Please review the comments in the code

File: library/mr_provisioner_preseed.py (hunk 2 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: Because we are talking about a file that is being passed as argument to the module, and to keep consistency with the behaviour of the kernel and initrd upload.
It could be translated when instanciating the object, but to keep it organized in my head I kept it at '/dev/null'.
If an empty string is clearer for everyone, I'll change all the '/dev/null' in PreseedUploader.

File: library/mr_provisioner_preseed.py (hunk 3 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: Because we are talking about a file that is being passed as argument to the module, and to keep consistency with the behaviour of the kernel and initrd upload.
It could be translated when instanciating the object, but to keep it organized in my head I kept it at '/dev/null'.
If an empty string is clearer for everyone, I'll change all the '/dev/null' in PreseedUploader.

File: tasks/provision.yml (hunk 6 above)
Comment (BaptisteGerondeau) [reply tasks/provision.yml]: Because we are talking about a file that is being passed as argument to the module.

File: library/mr_provisioner_preseed.py (hunk 5 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: Nope that must be me typing by hand, will remove.

File: library/mr_provisioner_preseed.py (hunk 4 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: Will correct.

File: library/mr_provisioner_preseed.py (hunk 1 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: Cheers, will correct

File: library/mr_provisioner_preseed.py (hunk 2 above)
Comment (niedbalski) [reply library/mr_provisioner_preseed.py]: I got your idea, but I stick to empty string.

File: library/mr_provisioner_preseed.py [hunk 7]
```diff
@@ -77,72 +80,149 @@
...
+        return False
+
+    def _get_preseed_from_file(self):
+        json_preseed = {}
+        contents = ''
+
+        with open(self.file, 'r') as fd:
+            lines = fd.readlines()
+        for line in lines:
+            contents += line
```
Comment (niedbalski) [inline library/mr_provisioner_preseed.py]: not required, json_preseed['content'] = lines

**Review by niedbalski — CHANGES_REQUESTED** [review]
Minor nitpicks

File: library/mr_provisioner_preseed.py (hunk 7 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: Ohhhhh that's neat !!!
Sadly it gives me a HTTP 400 BAD REQUEST....

File: library/mr_provisioner_preseed.py (hunk 2 above)
Comment (BaptisteGerondeau) [reply library/mr_provisioner_preseed.py]: In the python code why not, but I stick to '/dev/null' in the YAML... right ?

File: library/mr_provisioner_preseed.py [hunk 8]
```diff
@@ -1,12 +1,12 @@
-#!/usr/bin/python
+#!/usr/bin/env python3
```
Comment (danrue) [inline library/mr_provisioner_preseed.py]: I don't think we're quite ready to require python3. If you'd like to propose this, please do so in another PR and apply the change to every module consistently.

File: library/mr_provisioner_preseed.py [hunk 9]
```diff
@@ -77,72 +80,150 @@
...
+        if self.id != -1 and self.file != '':     #Exists and file given
+            try:
+                res = self._modify_preseed()
+            except ProvisionerError as e:
+                res['error'] = str(e)
+            return res
+        elif self.file != '':        #Doesn't exist and file given
+            preseed = self._get_preseed_from_file()
+            url = urljoin(self.url, '/api/v1/preseed')
+            r = requests.post(url, headers=self.authhead, data=json.dumps(preseed))
```
Comment (danrue) [inline library/mr_provisioner_preseed.py]: Consider handling this case (the post) similarly to how you handle the modify case (with a put). There's a lot of duplication between the post and the put case. Also, if one of the cases is in a function, why is the other inline?

File: library/mr_provisioner_preseed.py [hunk 10]
```diff
@@ -77,72 +80,150 @@
...
+        fetch the thing via the regular call in the Ansible role"""
+    def __init__(self, mrp_url, mrp_token, preseed_file, preseed_name,
+                 preseed_type, preseed_desc='', preseed_knowngood=False,
+                 preseed_public=False):
+        self.url = mrp_url
+        self.authhead = {'Authorization': mrp_token}
+        self.file = preseed_file
+        self.name = preseed_name
+        self.type = preseed_type
+        self.id = -1
```
Comment (danrue) [inline library/mr_provisioner_preseed.py]: I would probably use None here. Special values should be avoided.

**Review by danrue — CHANGES_REQUESTED** [review]
Overall this looks like a good improvement.

There are opportunities to make the PreseedUploader class more simple, but it is up to you whether it is worth one more round of refactoring.

In any case, have a look at some of my inline feedback, and then please rebase this pull request and make it a single commit on top of the latest master branch.

**Review by danrue — APPROVED** [review]

**Comment by danrue** [comment]
Thanks Baptiste!
